starlark/eval_test.go

```diff
@@ -963,8 +971,168 @@ func TestCancel(t *testing.T) {
+  cancel.star:6:3: in g
+`
+		if !strings.Contains(evalErr.Backtrace(), expectedBacktace) || !regexp.MustCompile(`cancel\.star:(10:\d+|9:\d+): in h`).MatchString(evalErr.Backtrace()) {
+			t.Errorf("ExecFile: expected backtrace to include calls to f(), g(), and h(); got %s", evalErr.Backtrace())
+		}
+	}
+}
+
+func TestCancelWithError(t *testing.T) {
+	// A thread canceled before it begins executes no code.
+	{
+		thread := new(starlark.Thread)
+		cancelErr := errors.New("nope")
+		thread.CancelWithError(cancelErr)
+		_, err := starlark.ExecFile(thread, "precancel.star", `x = 1//0`, nil)
+		if fmt.Sprint(err) != "Starlark computation canceled: nope" {
+			t.Errorf("ExecFile returned error %q, want cancellation", err)
+		}
+		if !errors.Is(err, starlark.ErrCanceled) {
+			t.Errorf("ExecFile didn't return error starlark.ErrCanceled")
+		}
+		if !errors.Is(err, cancelErr) {
+			t.Errorf("ExecFile didn't return expected error")
+		}
+
+		// cancellation is sticky
+		_, err = starlark.ExecFile(thread, "precancel.star", `x = 1//0`, nil)
+		if fmt.Sprint(err) != "Starlark computation canceled: nope" {
+			t.Errorf("ExecFile returned error %q, want cancellation", err)
+		}
+		if !errors.Is(err, starlark.ErrCanceled) {
+			t.Errorf("ExecFile: !errors.Is(err, starlark.ErrCanceled); got %T, %v", err, err)
+		}
+		if !errors.Is(err, cancelErr) {
+			t.Errorf("ExecFile didn't return expected error")
+		}
+	}
+	// A thread canceled during a built-in executes no more code.
+	{
+		thread := new(starlark.Thread)
+		var cancelErr error
+		predeclared := starlark.StringDict{
+			"stopit": starlark.NewBuiltin("stopit", func(thread *starlark.Thread, b *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+				cancelErr = errors.New(fmt.Sprint(args[0]))
+				thread.CancelWithError(cancelErr)
+				return starlark.None, nil
+			}),
+		}
+		_, err := starlark.ExecFile(thread, "stopit.star", `msg = 'nope'; stopit(msg); x = 1//0`, predeclared)
+		if fmt.Sprint(err) != `Starlark computation canceled: "nope"` {
+			t.Errorf("ExecFile returned error %q, want cancellation", err)
+		}
+		if !errors.Is(err, starlark.ErrCanceled) {
+			t.Errorf("ExecFile: !errors.Is(err, starlark.ErrCanceled); got %T, %v", err, err)
+		}
+		if !errors.Is(err, cancelErr) {
+			t.Errorf("ExecFile didn't return expected error")
+		}
+	}
+	// An external cancelation returns a wrapped starlark.ErrCanceled and the
+	// expected backtrace.
+	{
+		thread := new(starlark.Thread)
+		var ready atomic.Bool
+		predeclared := starlark.StringDict{
+			"fibonacci": fib{},
+			"ready": starlark.NewBuiltin("ready", func(_ *starlark.Thread, _ *starlark.Builtin, _ starlark.Tuple, _ []starlark.Tuple) (starlark.Value, error) {
+				ready.Store(true)
+				return starlark.True, nil
+			}),
+		}
+		src := `
+def f():
+	g()
+
+def g():
+	h()
+
+def h():
+	for x in fibonacci:
+		ready()
+
+f()
+`
+		cancelErr := errors.New("nope")
+		go func() {
+			for !ready.Load() {
```

Comment:
This is not how to wait for an event; use a semaphore (sync.WaitGroup) or just a channel.

But why not just have the `ready` function cancel its own Thread? The concurrency of polling is not what's being tested here. The Starlark program then reduces to:
```
def f(): g()
def g(): h()
def h(): cancel()
```
and the cancel builtin function becomes:
```
"cancel": starlark.NewBuiltin(...) {
		thread.CancelWithError(cancelErr)
		return starlark.None, nil
}
```

Reply:
I see value in testing internally (sync?) and externally (async?) triggered cancellations, but I'm happy to change it.